### PR TITLE
Mmca 4773 : Upgrade scala & library dependancies to latest version

### PR DESCRIPTION
**Pull request by HariHmrc**

### CODE REVIEW

File: app/uk/gov/hmrc/customs/emailfrontend/config/ErrorHandler.scala [hunk 1]
```diff
@@ -22,12 +22,14 @@ import play.api.mvc.Request
...
-import uk.gov.hmrc.play.bootstrap.frontend.http.FrontendErrorHandler
+import uk.gov.hmrc.play.bootstrap.frontend.http.LegacyFrontendErrorHandler
+
+import scala.concurrent.{ExecutionContext, Future}
 
 @Singleton
 class ErrorHandler @Inject()(override val messagesApi: MessagesApi,
                              errorView: error_template,
-                             customErrorView: problem_with_this_service) extends FrontendErrorHandler {
+                             customErrorView: problem_with_this_service) extends LegacyFrontendErrorHandler {
```
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/customs/emailfrontend/config/ErrorHandler.scala:32]: Please make use of FrontendErrorHandler and make related changes

File: app/uk/gov/hmrc/customs/emailfrontend/connectors/Save4LaterConnector.scala [hunk 2]
```diff
@@ -26,89 +26,103 @@ import uk.gov.hmrc.customs.emailfrontend.connectors.http.responses.{BadRequest,
...
-class Save4LaterConnector @Inject()(http: HttpClient,
+class Save4LaterConnector @Inject()(http: HttpClientV2,
                                     appConfig: AppConfig,
                                     audit: Auditable)(implicit ec: ExecutionContext) extends Logging {
 
   def getEmailDetails(id: String, key: String)
                      (implicit hc: HeaderCarrier, reads: Reads[EmailDetails]): Future[Option[EmailDetails]] = {
 
-    val url = s"${appConfig.save4LaterUrl}/$id/$key"
+    val getUrl = s"${appConfig.save4LaterUrl}/$id/$key"
```
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/customs/emailfrontend/connectors/Save4LaterConnector.scala:43]: Better to rename it to endPointUrl or endPointUri or urlString as this name gives the feeling of method name.

Same for other methods of this class

File: app/uk/gov/hmrc/customs/emailfrontend/connectors/Save4LaterConnector.scala [hunk 3]
```diff
@@ -26,89 +26,103 @@ import uk.gov.hmrc.customs.emailfrontend.connectors.http.responses.{BadRequest,
...
-    logger.info(s"DELETE: $url")
-    auditCallRequest(url, JsNull)
-
-    http.DELETE[HttpResponse](url).map { response =>
-      auditCallResponse(url, response.status)
-      response.status match {
-        case NO_CONTENT => Right(())
-        case _ => Left(BadRequest)
+    val deleteUrl = s"${appConfig.save4LaterUrl}/$id"
+    logger.info(s"DELETE: $deleteUrl")
```
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/customs/emailfrontend/connectors/Save4LaterConnector.scala:111]: Better to give a line break before this line to separate logger from the url and to make it more readable

File: app/uk/gov/hmrc/customs/emailfrontend/connectors/Save4LaterConnector.scala [hunk 4]
```diff
@@ -26,89 +26,103 @@ import uk.gov.hmrc.customs.emailfrontend.connectors.http.responses.{BadRequest,
...
-    logger.info(s"PUT: $url")
-    auditCallRequest(url, payload)
-
-    http.PUT[JsValue, HttpResponse](url, payload).map { response =>
-      auditCallResponse(url, response.status)
-      response.status match {
-        case NO_CONTENT | CREATED | OK => Right(())
-        case _ => Left(BadRequest)
+    val putUrl = s"${appConfig.save4LaterUrl}/$id/$key"
+    logger.info(s"PUT: $putUrl")
```
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/customs/emailfrontend/connectors/Save4LaterConnector.scala:90]: Better to give a line break before this line to separate logger from the url and to make it more readable

File: app/uk/gov/hmrc/customs/emailfrontend/connectors/UpdateVerifiedEmailConnector.scala [hunk 5]
```diff
@@ -40,11 +41,12 @@ class UpdateVerifiedEmailConnector @Inject()(appConfig: AppConfig,
...
 
-    http.PUT[VerifiedEmailRequest, VerifiedEmailResponse](appConfig.updateVerifiedEmailUrl, request).map { resp =>
-      auditRequest(currentEmail, newEmail, eori, "changeEmailAddressConfirmed")
-      Right(resp)
-
-    } recover {
+    http.put(url"${appConfig.updateVerifiedEmailUrl}")
+      .withBody[VerifiedEmailRequest](request)
+      .execute[VerifiedEmailResponse].map { resp =>
+        auditRequest(currentEmail, newEmail, eori, "changeEmailAddressConfirmed")
```
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/customs/emailfrontend/connectors/UpdateVerifiedEmailConnector.scala:47]: Please format it from line 47-49 as it is not intended correctly

File: app/uk/gov/hmrc/customs/emailfrontend/views/Layout.scala.html
Comment (tyagikumarvineet) [inline app/uk/gov/hmrc/customs/emailfrontend/views/Layout.scala.html:1]: Please write the test cases for this view

File: project/AppDependencies.scala [hunk 6]
```diff
@@ -2,21 +2,37 @@ import sbt.*
...
-  private val bootstrapVersion = "8.6.0"
+  private val bootstrapVersion = "9.1.0"
 
   val compile: Seq[ModuleID] = Seq(
     play.sbt.PlayImport.ws,
     "uk.gov.hmrc" %% "bootstrap-frontend-play-30" % bootstrapVersion,
-    "uk.gov.hmrc" %% "play-frontend-hmrc-play-30" % "9.10.0",
-    "uk.gov.hmrc.mongo" %% "hmrc-mongo-play-30" % "1.9.0",
-    "uk.gov.hmrc" %% "emailaddress-play-30" % "4.0.0",
+    "uk.gov.hmrc" %% "play-frontend-hmrc-play-30" % "10.4.0",
```
Comment (tyagikumarvineet) [inline project/AppDependencies.scala:10]: Better to update it to 10.5.0 as it is the latest version, if you face any issues then stay with 10.4.0

File: project/AppDependencies.scala [hunk 7]
```diff
@@ -2,21 +2,37 @@ import sbt.*
...
+    "uk.gov.hmrc.mongo" %% "hmrc-mongo-play-30" % "2.1.0",
```
Comment (tyagikumarvineet) [inline project/AppDependencies.scala:11]: Better to update it to 2.2.0 as it is the latest version. Should you face the issue then stay with 2.1.0

File: test/uk/gov/hmrc/customs/emailfrontend/PackageSpec.scala [hunk 8]
```diff
@@ -16,11 +16,11 @@
 
 package uk.gov.hmrc.customs.emailfrontend
 
-import org.scalatest.matchers.must.Matchers.convertToAnyMustWrapper
 import uk.gov.hmrc.customs.emailfrontend.utils.SpecBase
 import uk.gov.hmrc.customs.emailfrontend.utils.Utils.{emptyString, singleSpace}
+import org.scalatest.matchers.must.Matchers.mustBe
 
-class PackageSpec extends SpecBase {
+class PackageSpec extends SpecBase{
```
Comment (tyagikumarvineet) [inline test/uk/gov/hmrc/customs/emailfrontend/PackageSpec.scala:23]: Not properly format, there should be a space between SpecBase and {

File: test/uk/gov/hmrc/customs/emailfrontend/connectors/CustomsDataStoreConnectorSpec.scala [hunk 9]
```diff
@@ -16,51 +16,56 @@
...
   private val testEori: Eori = Eori("GB1234556789")
   private val testEmail = "email@test.com"
   private val testDateTime = LocalDateTime.parse("2021-01-01T11:11:11.111Z", dateFormatter02)
-  private val requestBody: UpdateEmail = UpdateEmail(testEori, testEmail, testDateTime)
-  private val headers = Seq("Content-Type" -> "application/json")
 
   override def beforeEach(): Unit = {
-    reset(mockHttp, mockAuditable, mockAppConfig)
-    when(mockAppConfig.customsDataStoreUrl).thenReturn(url)
+    reset(mockHttp, mockAuditable, mockAppConfig, requestBuilder)
```
Comment (tyagikumarvineet) [inline test/uk/gov/hmrc/customs/emailfrontend/connectors/CustomsDataStoreConnectorSpec.scala:55]: Please add a line break after this line to make it more readable and to separate from the when statements

File: test/uk/gov/hmrc/customs/emailfrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 10]
```diff
@@ -16,36 +16,40 @@
...
+  private val mockHttpClient = mock[HttpClientV2]
+
+  private val emailBaseUrl = "http://localhost:9744/email-verification"
 
   val connector = new EmailVerificationConnector(mockHttpClient, mockAppConfig, mockAuditable)
   implicit val hc: HeaderCarrier = HeaderCarrier()
 
   before {
-    reset(mockAuditable, mockAppConfig, mockHttpClient)
+    reset(mockAuditable, mockAppConfig, mockHttpClient, requestBuilder)
```
Comment (tyagikumarvineet) [inline test/uk/gov/hmrc/customs/emailfrontend/connectors/EmailVerificationConnectorSpec.scala:49]: Add a line break after this line to make it more readable

File: test/uk/gov/hmrc/customs/emailfrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 11]
```diff
@@ -89,15 +87,11 @@ class EmailVerificationConnectorSpec extends SpecBase with BeforeAndAfter {
...
-        when(mockHttpClient.POST[JsObject, EmailVerificationStateResponse](any,
-          any,
-          any)(
-          any,
-          any,
-          any[HeaderCarrier],
-          any)
-        ).thenReturn(Future.successful(Left(
-          EmailVerificationStateErrorResponse(INTERNAL_SERVER_ERROR, "Internal Server Error"))))
+        when(requestBuilder.withBody(any())(any(), any(), any())).thenReturn(requestBuilder)
```
Comment (tyagikumarvineet) [inline test/uk/gov/hmrc/customs/emailfrontend/connectors/EmailVerificationConnectorSpec.scala:90]: Add a line break before this line to make it more readable

File: test/uk/gov/hmrc/customs/emailfrontend/connectors/EmailVerificationConnectorSpec.scala [hunk 12]
```diff
@@ -110,21 +104,10 @@ class EmailVerificationConnectorSpec extends SpecBase with BeforeAndAfter {
...
-                "email" -> "email-address",
-                "templateId" -> "verifyEmailAddress",
-                "templateParameters" -> Json.obj(),
-                "linkExpiryDuration" -> "P3D",
-                "continueUrl" -> "test-continue-url")
-            ),
-            any
-          )(any, any, any[HeaderCarrier], any)
-        ).thenReturn(Future.successful(Right(EmailVerificationRequestSent)))
+        when(requestBuilder.withBody(any())(any(), any(), any())).thenReturn(requestBuilder)
```
Comment (tyagikumarvineet) [inline test/uk/gov/hmrc/customs/emailfrontend/connectors/EmailVerificationConnectorSpec.scala:107]: Add a line break after this line to make it more readable

File: test/uk/gov/hmrc/customs/emailfrontend/connectors/Save4LaterConnectorSpec.scala [hunk 13]
```diff
@@ -16,23 +16,27 @@
...
+import uk.gov.hmrc.http.client.{HttpClientV2, RequestBuilder}
+import org.mockito.ArgumentMatchers.any
 
-import scala.concurrent.Future
+import scala.concurrent.{ExecutionContext, Future}
 
 class Save4LaterConnectorSpec extends SpecBase {
 
-  val mockHttpClient = mock[HttpClient]
+  val mockHttpClient = mock[HttpClientV2]
```
Comment (tyagikumarvineet) [inline test/uk/gov/hmrc/customs/emailfrontend/connectors/Save4LaterConnectorSpec.scala:38]: Better to add the return type for this and for line no 39 as these are public members

File: test/uk/gov/hmrc/customs/emailfrontend/connectors/Save4LaterConnectorSpec.scala [hunk 14]
```diff
@@ -226,9 +246,10 @@ class Save4LaterConnectorSpec extends SpecBase {
 
       val badRequestException = new BadRequestException("testMessage")
       val testJson = Json.toJson("test")
-
-      when(mockHttpClient.PUT[JsValue, HttpResponse](any, any, any)(any, any, any, any)
-      ).thenReturn(Future.failed(badRequestException))
+      when(requestBuilder.withBody(any())(any(), any(), any())).thenReturn(requestBuilder)
```
Comment (tyagikumarvineet) [inline test/uk/gov/hmrc/customs/emailfrontend/connectors/Save4LaterConnectorSpec.scala:249]: Add a line before after this line to make it more readable

File: test/uk/gov/hmrc/customs/emailfrontend/controllers/EmailLanguageControllerSpec.scala [hunk 15]
```diff
@@ -21,6 +21,7 @@ import play.api.i18n.Lang
 import play.api.test.Helpers.running
 import uk.gov.hmrc.customs.emailfrontend.utils.{FakeIdentifierAgentAction, SpecBase}
 
+
```
Comment (tyagikumarvineet) [inline test/uk/gov/hmrc/customs/emailfrontend/controllers/EmailLanguageControllerSpec.scala:24]: Remove this extra line break

**Comment by tyagikumarvineet** [comment]
Please address the below compilation warning

- [E121] Pattern Match Warning: /Users/vineet.tyagi/hmrc/codebase/customs-email-frontend/app/uk/gov/hmrc/customs/emailfrontend/controllers/WhatIsYourEmailController.scala:140:15
[warn] 140 |          case _ => Redirect(routes.WhatIsYourEmailController.problemWithService())
[warn]     |               ^
[warn]     |Unreachable case except for null (if this is intentional, consider writing case null => instead).

File: app/uk/gov/hmrc/customs/emailfrontend/connectors/UpdateVerifiedEmailConnector.scala (hunk 5 above)
Comment (HariHmrc) [reply app/uk/gov/hmrc/customs/emailfrontend/connectors/UpdateVerifiedEmailConnector.scala:47]: even after formatting in intellij, it looks the same.. so leaving it unfixed

File: test/uk/gov/hmrc/customs/emailfrontend/views/LayoutSpec.scala [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+import play.api.mvc.AnyContentAsEmpty
+import play.api.test.FakeRequest
+import play.twirl.api.Html
+import uk.gov.hmrc.customs.emailfrontend.utils.{FakeIdentifierAgentAction, SpecBase}
+import uk.gov.hmrc.customs.emailfrontend.utils.Utils.emptyString
+import uk.gov.hmrc.customs.emailfrontend.views.html.Layout
+import org.scalatest.matchers.must.Matchers.mustBe
+import uk.gov.hmrc.customs.emailfrontend.config.AppConfig
+
+
```
Comment (tyagikumarvineet) [inline test/uk/gov/hmrc/customs/emailfrontend/views/LayoutSpec.scala:32]: Remove this extra line break

**Review by tyagikumarvineet — APPROVED** [review]

**Pull request merged**